### PR TITLE
added code to fetch data from dal.py

### PR DESCRIPTION
**Pull request by mudit1990**
@doodhwala @sarankrishna920 please review
Code to fetch data from dal.py
-- to avoid merge conflicts this needs to be merged before you make any changes [file name changed]

** load-from_db contains code to load from db and dump it as csv
** __db_query_all and __db_query_partial contain the actual queries
** __clean_data cleans the data (handles ',' '\n' and '\t' within the columns) before dumping it as csv
** a strict file naming convention is followed which is used for reading files
** get_emails fetches the data from a dumped csv by selecting the most appropriate file that meets the requirement
** files are stored and read from the data folder

### CODE REVIEW

File: code/dal.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+def __db_query_all(db_conn):
+    """
+    :param db_conn:
+    :return: tuples of tuples containing data
+    """
+    query = "select message.mid, sender, group_concat(rvalue) as receivers, body from" + \
+            "EnronAHS.message inner join EnronAHS.recipientinfo on message.mid=recipientinfo.mid group by message.mid;"
+    cur = db_conn.cursor()
+    cur.execute("SET SESSION group_concat_max_len = 100000;")
```
Comment (mohit-surana) [inline code/dal.py:100]: Can you explain what is happening here?

File: code/dal.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    if rec_rows == -1:
+        rec_rows = max_rows
+    return str(rec_rows) + '_emails.csv'
+
+
+def __clean_data(r):
+    """
+    The function assumes an input as a tuple of tuples, cleans the data and returns a numpy array
+    """
+    data = np.asarray(r)
```
Comment (mohit-surana) [inline code/dal.py:80]: Why have we picked just these four characters?
Do you want to look at something along the lines of stopword removal?
Take a look at [this](https://stackoverflow.com/questions/19130512/stopword-removal-with-nltk) and [this](https://pythonspot.com/nltk-stop-words/)

I'm just laying it down here as possible enhancements for the future.

File: code/dal.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        file_name = 'all_emails.csv'
+    else:
+        file_name = str(num_emails) + '_emails.csv'
+        res = __db_query_partial(db_conn, num_emails)
+    data = __clean_data(res)
+    data = data[:,1:] # remove the first column that contains the random id
+    np.savetxt(file_path+file_name, data, fmt='%s', delimiter=',')
+
+
+def __get_appr_filename(num_emails, file_path):
```
Comment (mohit-surana) [inline code/dal.py:48]: Smart idea.

File: code/dal.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+
+def load_from_db(num_emails=100, fetch_all=False):
+    """loads the requested number of emails from the database and dumps the file as csv. For this function to
+    run the mysql database must be set up locally.
+    if fetch_all is false, file is stored as <num_emails>_emails.csv
+    if the fetch_all is true, file is stored as all_emails.csv
+    :param num_emails: The number of emails to be fetched from the db
+    :param fetch_all: Will fetch all the emails from db. Large join query that may take some time"""
+    db_conn = MySQLdb.connect(host='localhost', user='root', passwd='mudit1990', db='EnronAHS')
```
Comment (mohit-surana) [inline code/dal.py:35]: This is good enough for now.
In the future we can pull these from a constants file (gitignored)

File: code/dal.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+def get_emails(num_emails=100, fetch_all=False):
+    """ Returns requested number of emails from a csv file. The function returns a numpy array of dimension (N,3)
+    where N is the number of rows, dimension1 is senderId, dimension2 is receiverIds, dimension3 is message body
+    If there are multiple csv files present, it selects the appropriate file that meets the <num_emails> requirement
+    and returns the requested number of rows
+    :param num_emails: The number of rows to be returned
+    :param fetch_all: Ignores num_emails & fetches all the emails in the enron dataset. The number of returned rows may
+    be very large """
+    file_path = '../data/'
+    if fetch_all:
```
Comment (mohit-surana) [inline code/dal.py:15]: Good that this has been written in a forward looking fashion :)

**Review by mohit-surana — APPROVED** [review]
Looks good and others have been unblocked!
I've laid down a few minor details that can be taken care of as part of future enhancements.

:shipit:

File: code/dal.py (hunk 1 above)
Comment (mudit1990) [reply code/dal.py:100]: so basically i'm concatenating the receivers. max limit for group_concat is 1024. That was not enough so I set it higher. I'll put in an explicit comment

File: code/dal.py (hunk 2 above)
Comment (mudit1990) [reply code/dal.py:80]: yes, this is just preliminary. This layer will support generic cleaning that all model may require including stop words. As of now I just cleaned stuff that would prevent csv read and writes like '\n' and ','

File: code/dal.py (hunk 4 above)
Comment (mudit1990) [reply code/dal.py:35]: agreed

**Comment by doodhwala** [comment]
Closes #3

**Pull request merged**